File: SignalServiceKit/Storage/YDBStorage.m
//
// Copyright 2021 Signal Messenger, LLC
// SPDX-License-Identifier: AGPL-3.0-only
//

#import "YDBStorage.h"
#import "OWSFileSystem.h"
#import <SignalServiceKit/SignalServiceKit-Swift.h>

NS_ASSUME_NONNULL_BEGIN

#pragma mark -

@implementation YDBStorage

+ (NSString *)legacyDatabaseDirPath
{
    return [OWSFileSystem appDocumentDirectoryPath];
}

+ (NSString *)sharedDataDatabaseDirPath
{
    return [[OWSFileSystem appSharedDataDirectoryPath] stringByAppendingPathComponent:@"database"];
}

+ (NSString *)databaseFilename
{
    return @"Signal.sqlite";
}

+ (NSString *)databaseFilename_SHM
{
    return [self.databaseFilename stringByAppendingString:@"-shm"];
}

+ (NSString *)databaseFilename_WAL
{
    return [self.databaseFilename stringByAppendingString:@"-wal"];
}

+ (NSString *)legacyDatabaseFilePath
{
    return [self.legacyDatabaseDirPath stringByAppendingPathComponent:self.databaseFilename];
}

+ (NSString *)legacyDatabaseFilePath_SHM
{
    return [self.legacyDatabaseDirPath stringByAppendingPathComponent:self.databaseFilename_SHM];
}

+ (NSString *)legacyDatabaseFilePath_WAL
{
    return [self.legacyDatabaseDirPath stringByAppendingPathComponent:self.databaseFilename_WAL];
}

+ (NSString *)sharedDataDatabaseFilePath
{
    return [self.sharedDataDatabaseDirPath stringByAppendingPathComponent:self.databaseFilename];
}

+ (NSString *)sharedDataDatabaseFilePath_SHM
{
    return [self.sharedDataDatabaseDirPath stringByAppendingPathComponent:self.databaseFilename_SHM];
}

+ (NSString *)sharedDataDatabaseFilePath_WAL
{
    return [self.sharedDataDatabaseDirPath stringByAppendingPathComponent:self.databaseFilename_WAL];
}

#pragma mark -

+ (void)deleteYDBStorage
{
    [self deleteDatabaseFiles];
}

+ (void)deleteDatabaseFiles
{
    [OWSFileSystem deleteFileIfExists:self.legacyDatabaseFilePath];
    [OWSFileSystem deleteFileIfExists:self.legacyDatabaseFilePath_SHM];
    [OWSFileSystem deleteFileIfExists:self.legacyDatabaseFilePath_WAL];
    [OWSFileSystem deleteFileIfExists:self.sharedDataDatabaseFilePath];
    [OWSFileSystem deleteFileIfExists:self.sharedDataDatabaseFilePath_SHM];
    [OWSFileSystem deleteFileIfExists:self.sharedDataDatabaseFilePath_WAL];
    // NOTE: It's NOT safe to delete OWSPrimaryStorage.legacyDatabaseDirPath
    //       which is the app document dir.
    [OWSFileSystem deleteContentsOfDirectory:self.sharedDataDatabaseDirPath];
}


@end

NS_ASSUME_NONNULL_END

File: SignalServiceKit/Storage/YDBStorage.swift
//
// Copyright 2024 Signal Messenger, LLC
// SPDX-License-Identifier: AGPL-3.0-only
//

public enum YDBStorage {
    private static var legacyDatabaseDirPath: String {
        OWSFileSystem.appDocumentDirectoryPath()
    }
    private static var sharedDataDatabaseDirPath: String {
        OWSFileSystem.appSharedDataDirectoryPath().appendingPathComponent("database")
    }
    private static let databaseFilename: String = "Signal.sqlite"
    private static let databaseFilename_SHM: String = "\(databaseFilename)-shm"
    private static let databaseFilename_WAL: String = "\(databaseFilename)-wal"
    private static var legacyDatabaseFilePath: String {
        legacyDatabaseDirPath.appendingPathComponent(databaseFilename)
    }
    private static var legacyDatabaseFilePath_SHM: String {
        legacyDatabaseDirPath.appendingPathComponent(databaseFilename_SHM)
    }
    private static var legacyDatabaseFilePath_WAL: String {
        legacyDatabaseDirPath.appendingPathComponent(databaseFilename_WAL)
    }
    private static var sharedDataDatabaseFilePath: String {
        sharedDataDatabaseDirPath.appendingPathComponent(databaseFilename)
    }
    private static var sharedDataDatabaseFilePath_SHM: String {
        sharedDataDatabaseDirPath.appendingPathComponent(databaseFilename_SHM)
    }
    private static var sharedDataDatabaseFilePath_WAL: String {
        sharedDataDatabaseDirPath.appendingPathComponent(databaseFilename_WAL)
    }

    public static func deleteYDBStorage() {
        OWSFileSystem.deleteFileIfExists(legacyDatabaseFilePath)
        OWSFileSystem.deleteFileIfExists(legacyDatabaseFilePath_SHM)
        OWSFileSystem.deleteFileIfExists(legacyDatabaseFilePath_WAL)
        OWSFileSystem.deleteFileIfExists(sharedDataDatabaseFilePath)
        OWSFileSystem.deleteFileIfExists(sharedDataDatabaseFilePath_SHM)
        OWSFileSystem.deleteFileIfExists(sharedDataDatabaseFilePath_WAL)
        // NOTE: It's NOT safe to delete OWSPrimaryStorage.legacyDatabaseDirPath
        //       which is the app document dir.
        OWSFileSystem.deleteContents(ofDirectory: sharedDataDatabaseDirPath)
    }
}
